### PR TITLE
Feature StorageCapacityTracking for Powermax

### PR DESCRIPTION
**Pull request by tssushma**
# Description
Support storagecapacitytracking feature for PowerMax.

Additional context
This feature helps the scheduler to schedule the pod on a node (satisfying the topology constraints) only if the requested capacity is available on the storage array. 

# GitHub Issues
List the GitHub issues impacted by this PR:

| GitHub Issue # |
| -------------- |
|https://github.com/dell/csm/issues/876|

# Checklist:

- [x] I have performed a self-review of my own code to ensure there are no formatting, vetting, linting, or security issues
- [x] I have verified that new and existing unit tests pass locally with my changes
- [x] I have not allowed coverage numbers to degenerate
- [ ] I have maintained at least 90% code coverage
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have maintained backward compatibility

# How Has This Been Tested?
Please describe the tests that you ran to verify your changes. Please also list any relevant details for your test configuration
image

- [ ] Test A
- [ ] Test B

### CODE REVIEW

File: pkg/drivers/powermax.go [hunk 1]
```diff
@@ -53,6 +53,8 @@ const (
 	CSIPmaxVsphereHostname = "<X_CSI_VSPHERE_HOSTNAME>"
 	CSIPmaxVsphereHost     = "<X_CSI_VCENTER_HOST>"
 	CSIPmaxChap            = "<X_CSI_POWERMAX_ISCSI_ENABLE_CHAP>"
+	// CSIStorageCapacityEnabled - Storage capacity flag
+	CSIStorageCapacityEnabled = "false"
```
[the remark this answers is not in the document]
Comment (tssushma) [reply pkg/drivers/powermax.go]: This parameter is a redundant entry.
Rectified it and removed this from constant list.

**Review by delldubey — APPROVED** [review]
lgtm

**Comment by suryagupta4** [comment]
We can also give users an option for configuring the `capacity-poll-interval`
This is done for PowerStore & PowerScale, refer here:
https://github.com/dell/csm-operator/blob/main/samples/storage_csm_powerstore_v270.yaml#L65
https://github.com/dell/csm-operator/blob/main/samples/storage_csm_powerscale_v270.yaml#L249

**Comment by rensyct** [comment]
@tssushma , is this change intended for 2.7 or 2.8 version of the driver

**Comment by tssushma** [comment]
> @tssushma , is this change intended for 2.7 or 2.8 version of the driver



> @tssushma , is this change intended for 2.7 or 2.8 version of the driver

@rensyct , Done updating the version to v2.8.
@suryagupta4, Done adding capacity-poll-interval configuration variable for giving the users an option for configuring poll interval.

File: operatorconfig/driverconfig/powermax/v2.8.0/upgrade-path.yaml [hunk 2]
```diff
@@ -0,0 +1 @@
+minUpgradePath: v2.6.0
```
Comment (rensyct) [inline operatorconfig/driverconfig/powermax/v2.8.0/upgrade-path.yaml:1]: Should be 2.7.0

File: samples/storage_csm_powermax_v280.yaml [hunk 3]
```diff
@@ -0,0 +1,324 @@
...
+    # Default value: None
+    replicas: 2
+    # Default credential secret for Powermax, if not set it to ""
+    authSecret: powermax-creds
+    dnsPolicy: ClusterFirstWithHostNet
+    forceUpdate: false
+    forceRemoveDriver: true
+    common:
+      # Image for CSI PowerMax driver v2.7.0
+      image: dellemc/csi-powermax:v2.7.0
```
Comment (rensyct) [inline samples/storage_csm_powermax_v280.yaml:48]: should be v2.8.0

File: tests/config/driverconfig/powermax/v2.7.0/upgrade-path.yaml [hunk 4]
```diff
@@ -1 +1 @@
-minUpgradePath: v2.6.0
```
Comment (rensyct) [inline tests/config/driverconfig/powermax/v2.7.0/upgrade-path.yaml:1]: Please remove test files related to v2.7.0

File: tests/config/driverconfig/powermax/v2.8.0/upgrade-path.yaml [hunk 5]
```diff
@@ -0,0 +1 @@
+minUpgradePath: v2.6.0
```
Comment (rensyct) [inline tests/config/driverconfig/powermax/v2.8.0/upgrade-path.yaml:1]: Should be v2.7.0

File: operatorconfig/driverconfig/powermax/v2.8.0/controller.yaml [hunk 6]
```diff
@@ -0,0 +1,328 @@
...
+            - "--leader-election"
+            - "--snapshot-name-uuid-length=10"
+          env:
+            - name: ADDRESS
+              value: /var/run/csi/csi.sock
+          volumeMounts:
+            - name: socket-dir
+              mountPath: /var/run/csi
+        - name: driver
+          image: dellemc/csi-powermax:v2.7.0
```
Comment (rensyct) [inline operatorconfig/driverconfig/powermax/v2.8.0/controller.yaml:233]: should be v2.8.0

File: operatorconfig/driverconfig/powermax/v2.8.0/node.yaml [hunk 7]
```diff
@@ -0,0 +1,256 @@
...
+      dnsPolicy: ClusterFirstWithHostNet
+      containers:
+        - name: driver
+          command: ["/csi-powermax.sh"]
+          securityContext:
+            privileged: true
+            capabilities:
+              add: ["SYS_ADMIN"]
+            allowPrivilegeEscalation: true
+          image: dellemc/csi-powermax:v2.7.0
```
Comment (rensyct) [inline operatorconfig/driverconfig/powermax/v2.8.0/node.yaml:90]: Should be v2.8.0

File: samples/storage_csm_powermax_v280.yaml [hunk 8]
```diff
@@ -0,0 +1,324 @@
...
+      # fsGroupPolicy: Defines if the underlying volume supports changing ownership and permission of the volume before being mounted.
+      # Allowed values: ReadWriteOnceWithFSType, File , None
+      # Default value: ReadWriteOnceWithFSType
+      fSGroupPolicy: "ReadWriteOnceWithFSType"
+      # storageCapacity: Helps the scheduler to schedule the pod on a node satisfying the topology constraints, only if the requested capacity is available on the storage array
+      # Allowed values:
+      #   true: enable storage capacity tracking
+      #   false: disable storage capacity tracking
+      storageCapacity: true
+    # Config version for CSI PowerMax v2.7.0 driver
```
Comment (rensyct) [inline samples/storage_csm_powermax_v280.yaml:34]: Should be v2.8.0

File: samples/storage_csm_powermax_v280.yaml [hunk 9]
```diff
@@ -0,0 +1,324 @@
...
+    # Allowed values: n, where n > 0
+    # Default value: None
+    replicas: 2
+    # Default credential secret for Powermax, if not set it to ""
+    authSecret: powermax-creds
+    dnsPolicy: ClusterFirstWithHostNet
+    forceUpdate: false
+    forceRemoveDriver: true
+    common:
+      # Image for CSI PowerMax driver v2.7.0
```
Comment (rensyct) [inline samples/storage_csm_powermax_v280.yaml:47]: Should be v2.8.0

File: tests/config/driverconfig/powermax/v2.8.0/node.yaml [hunk 10]
```diff
@@ -0,0 +1,254 @@
...
+      dnsPolicy: ClusterFirstWithHostNet
+      containers:
+        - name: driver
+          command: ["/csi-powermax.sh"]
+          securityContext:
+            privileged: true
+            capabilities:
+              add: ["SYS_ADMIN"]
+            allowPrivilegeEscalation: true
+          image: dellemc/csi-powermax:v2.7.0
```
Comment (rensyct) [inline tests/config/driverconfig/powermax/v2.8.0/node.yaml:90]: Should be v2.8.0

**Review by rensyct — CHANGES_REQUESTED** [review]

**Review by shanmydell — APPROVED** [review]

**Pull request merged**